framework/include/interfaces/DiffusionHDGInterface.h

```diff
@@ -0,0 +1,476 @@
+                             const MooseArray<Gradient> & vector_sol,
+                             const MooseArray<Number> & scalar_sol,
+                             const MooseArray<Number> & lm_sol);
+
+  template <typename DiffusionHDG>
+  static void lmFaceJacobian(DiffusionHDG & obj,
+                             const unsigned int i_offset,
+                             const unsigned int vector_j_offset,
+                             const unsigned int scalar_j_offset,
+                             const unsigned int lm_j_offset);
+
+  template <typename DiffusionHDG>
+  static void vectorDirichletResidual(DiffusionHDG & obj,
+                                      const unsigned int i_offset,
+                                      const Function & dirichlet_function);
+
+  template <typename DiffusionHDG>
+  static void scalarDirichletResidual(DiffusionHDG & obj,
+                                      const unsigned int i_offset,
+                                      const MooseArray<Gradient> & vector_sol,
+                                      const MooseArray<Number> & scalar_sol,
+                                      const Function & dirichlet_function);
+
+  template <typename DiffusionHDG>
+  static void scalarDirichletJacobian(DiffusionHDG & obj,
+                                      const unsigned int i_offset,
+                                      const unsigned int vector_j_offset,
+                                      const unsigned int scalar_j_offset);
+
+  const MooseVariableFE<Real> & _u_var;
+  const MooseVariableFE<RealVectorValue> & _grad_u_var;
+  const MooseVariableFE<Real> & _u_face_var;
+
+  // Containers for dof indices
+  const std::vector<dof_id_type> & _qu_dof_indices;
+  const std::vector<dof_id_type> & _u_dof_indices;
+  const std::vector<dof_id_type> & _lm_u_dof_indices;
+
+  // local solutions at quadrature points
+  const MooseArray<Gradient> & _qu_sol;
+  const MooseArray<Number> & _u_sol;
+  const MooseArray<Number> & _lm_u_sol;
+
+  // Element shape functions
+  const MooseArray<std::vector<RealVectorValue>> & _vector_phi;
+  const MooseArray<std::vector<Real>> & _scalar_phi;
+  const MooseArray<std::vector<RealVectorValue>> & _grad_scalar_phi;
+  const MooseArray<std::vector<Real>> & _div_vector_phi;
+
+  // Face shape functions
+  const MooseArray<std::vector<RealVectorValue>> & _vector_phi_face;
+  const MooseArray<std::vector<Real>> & _scalar_phi_face;
+  const MooseArray<std::vector<Real>> & _lm_phi_face;
+
+  // Material properties
+  const MaterialProperty<Real> & _diff;
+
+  // Number of dofs on elem
+  std::size_t _vector_n_dofs;
+  std::size_t _scalar_n_dofs;
+  std::size_t _lm_n_dofs;
+
+  /// Local sizes of the systems
+  std::size_t _primal_size, _lm_size;
+
+  /// Our stabilization coefficient
+  static constexpr Real _tau = 1;
+};
+
+template <typename DiffusionHDG>
+void
+DiffusionHDGInterface::resizeData(DiffusionHDG & obj)
+{
+  obj._vector_n_dofs = obj._qu_dof_indices.size();
```

Comment:
It does allow for this base "interface" to access protected/private members of derived classes

Reply:
because of the diamond problem I cannot have both `NavierStokesHDGAssemblyHelper` inherit from `DiffusionHDGAssemblyHelper` and `NavierStokesHDGKernel` inherit from `DiffusionHDGKernel`. I'm pretty sure that's the core reason of why things end up looking a bit weird